Add MCL support

ThePython10110 opened the pull request:
This PR adds MineClone2/Mineclonia support. So far, I've done the itemstrings/crafting and sounds, but still need to update formspecs and a few other things (including node breakability). Most things seem to work, although I haven't tested everything.

ZenonSeth commented:
Hey, thanks for your PR, and nice work! I saw the issue you raised a couple of months ago but sadly due to real life issues I've had no time to work on Logistica, and I also have never played any Mineclone or related derivatives. I can test this out this weekend and I can also work on this on top of your PR.

At a first glance, it looks pretty good - I have two minor points:

- One minor and quick issue is importing `logistica.itemstrings` as `i` in all those files - in general I try to avoid using one letter tables in prod code (I do have one for debug only, but it's nowhere in production). Specificall `i` is a very common letter to be used for variables, so chances of overlap are great, so something like `local items = logistica.itemstrings` would be better. 
- Second thing: specifically on how things are declared in `compat_mcl.lua` - the problem with declaring every mod as an optional dependency is that you can't guarantee they'll be there. The way `compat_mcl` is written, if a user doesn't have neither `mcl_core` nor `default` then some code will crash when trying to refer to `default`, since there's no checks that `default` exists. Ideally the mod should never crash, and provide a meaningful message on startup to indicate that it requires **either** mcl_core **or*** default + bucket. I can look into improving this if you are going to look at other things, like formspec.

I'm curious, since I don't play MCL/etc.  - what exactly needs to be updated for node breakability? Do MCL nodes not use the standard tool-based breakability (aka 'choppy', 'cracky', 'oddly_breakable_by_hand` ) ?

ThePython10110 commented:
1. I didn't think about `i` being so common, but it definitely is, so I'll change it.
2. I will add a check for this.

MCL uses its own system to emulate Minecraft's breaking times. Basically, each node will have groups for tools it can be broken with (`pickaxey`, `handy`, `axey`, etc.), and each node also has its own hardness value, specified with `_mcl_hardness` in the definition. It's a bit more complicated than that, but that's basically how it works.

ZenonSeth commented on util/compat_mcl.lua:
```diff
@@ -0,0 +1,37 @@
+-- Returns a player's inventory formspec with the correct width and hotbar position for the current game
+function logistica.inventory_formspec(x,y)
+    local formspec
+    if exchangeclone.mcl then
```
```suggestion
    if mcl then
```

ThePython10110 replied:
Whoops. I copied that from one of my mods.

ZenonSeth commented on util/compat_mcl.lua:
```diff
@@ -0,0 +1,37 @@
+            "list[current_player;main;"..x..","..(y+4)..";9,1]"
+    else
+        formspec = "list[current_player;main;"..x..","..y..";8,1]"..
+        "list[current_player;main;"..x..","..(y+1.5)..";8,3;8]"
```
```suggestion
        "list[current_player;main;"..x..","..(y + 1.25)..";8,3;8]"
```
1.25 seems to be MC's inventory slot spacing, this makes it look seamless

ThePython10110 replied:
Both MTG and MCL generally have a slightly larger gap between the hotbar and the rest of the inventory. That's why I did it the way I did.

ThePython10110 commented:
This is just because I decided to add the extra bit of separation between the hotbar and the rest of the inventory. I accidentally added 0.5 instead of 0.25 (and in some cases, I somehow did math wrong and added even more, such as the cobble gen). I'm working on that right now.

ZenonSeth commented on logic/access_point_formspec.lua:
```diff
@@ -189,6 +189,15 @@ local function get_access_point_formspec(pos, invName, optMeta, playerName)
+    local player = minetest.get_player_by_name(playerName)
+    if not player then return end
+    local inv = player:get_inventory()
+    if inv then
+      inv:set_width("craft", 3)
+      inv:set_size("craft", 9)
+    end
```
This is permanently going to change the MCL inv size isn't it? Any negative side-effects from that? Your main inv crafting gird in MCL only shows 2x2 right? so the items in the other 5 slots will still be there but hidden?

ThePython10110 replied:
No, I checked. Every time you open the normal inventory formspec, it sets the size to 2x2, and every time you open a crafting table, it sets it to 3x3. This is how MCL does it already.

Every time you exit the MCL crafting formspec, your crafting grid is emptied (either by returning items to your inventory or dropping them if there's not enough room). I suppose this is something that needs to be added.

ZenonSeth replied:
If it sets the size to 2x2, it means it will delete items that were left in the crafting grid - so yes, you would have to find a way to drop the items out of the Access Point. 

Do be aware that the Access Point UI is used by the Wireless Access pad, the exact same code is called as though the user right-clicked an Access Point (this is why I use detached inventories for most of it, and not node inventories). 

So if you're dropping items once the user closes an access point, they are best dropped at the user's position, rather than the access point's position.

ThePython10110 replied:
Just found this in MineClone2's code:
```
-- Drop items in craft grid and reset inventory on closing
minetest.register_on_player_receive_fields(function(player, formname, fields)
	if fields.quit then
		mcl_util.move_player_list(player, "craft")
		mcl_util.move_player_list(player, "enchanting_lapis")
		mcl_util.move_player_list(player, "enchanting_item")
		if not minetest.is_creative_enabled(player:get_player_name()) and (formname == "" or formname == "main") then
			set_inventory(player)
		end
	end
end)
```

So it looks like it clears the crafting grid upon closing *any* formspec, which is convenient.

ZenonSeth commented on registration/machines_api_reg.lua:
```diff
@@ -202,9 +206,9 @@ logistica.register_requester("Bulk Request Inserter\nInserts up to 64 items at a
-logistica.register_reservoir("lava", "Lava", "bucket:bucket_lava", "default_lava.png", 8)
-logistica.register_reservoir("water", "Water", "bucket:bucket_water", "default_water.png")
-logistica.register_reservoir("river_water", "River Water", "bucket:bucket_river_water", "default_river_water.png")
+logistica.register_reservoir("lava", "Lava", itemstrings.lava_bucket, "default_lava.png", 8)
+logistica.register_reservoir("water", "Water", itemstrings.water_bucket, "default_water.png")
+logistica.register_reservoir("river_water", "River Water", itemstrings.river_water_bucket, "default_river_water.png")
```
I think the textures (the 4th parameter) need to be updated, maybe need another table in compat_mcl like `logistica.base_textures` or something - because I doubt MCL also uses `default_water.png` right?

ThePython10110 replied:
Right. I haven't gotten that far yet, since the only water/lava textures MCL has are animated (although there's probably a texture modifier that grabs a specific part of the texture, but I just haven't checked yet).

ZenonSeth replied:
Yeah, I downloaded MCL2, using the animated texture doesn't really work unfortunately
![image](https://github.com/ZenonSeth/logistica/assets/118483769/f3ff0bba-aeb1-4a1e-ba3f-c98febecf127)

I don't think there's a crop modifier to create textures unfortunately - if that becomes an issue I can include some MCL specific textures in the modpack that match the default MCL looks. I know that isn't idea if there's texture packs, but these textures aren't shown that much, and as long as they are indicative of the liquid type, it should be fine.

ThePython10110 replied:
I found a solution: `^[sheet:1x16:0,0`

ThePython10110 commented:
I can't figure out why the autocrafter formspec has black text in MCL. I thought maybe adding `no_prepend[]` to the lava furnace background might fix it, but it didn't.

ThePython10110 commented:
Anything else I need to do? I haven't tested *everything* yet, but I'm working on it.

ZenonSeth commented:
> Anything else I need to do? I haven't tested _everything_ yet, but I'm working on it.

As far as MCL compat, I'm not sure, as I said, I have not played that game though I did download it, I will try to test it tomorrow. (sidenote: is there an equivalent in MCL of how unified_inv acts as a useful give inv for admin testing?). I think for most of MCL testing I would rely that you know what you're doing

What I want to do is double check all functionality on my MT test world for this on your fork, before merging this.

One thing I want to improve the formspec sizing to what I mentioned above - having a single function that takes a fixed size of w,h and adjusts it to MCL size where necessary. I am not a fan of the current way of randomly adding arbitrary numbers to the size. I will do that tomorrow as well and double check as much of it on MCL as I can.

ThePython10110 commented:
The only thing I did was add a 0.25 extra space between the hotbar and the rest of the inventory, and updated the formspec size to match (and I fixed the formspec size after you pointed out that it was too much).

ZenonSeth commented:
I went ahead and did the formspec changes now - see the commit I pushed - basically it's just to make reading the formspec sizes easier for me at least: rather than thinking why  4 or 2 is being arbitrarily added to the width, this makes the size be obvious at a glance and it still adjusts it the same way as you did for MCL.

ZenonSeth commented:
OK, tested it on both MTG and MCL, the UI looks good, basic functionality is there, I think this is good to go.

ZenonSeth commented:
@ThePython10110 So I spend an hour going through the MCL api docs and trying out a few things, but no success. The Reservoirs - all variants - empty/full and silverin/obsidian are not diggable in MCL in survival mode at all, even though inspecting the node shows they have the appropriate (as far as I can tell) groups set: 
![image](https://github.com/ZenonSeth/logistica/assets/118483769/3ef0619a-ae71-4c95-a544-7e87021c2a5f)

I don't have more time to look into it right now, will come back to it later - I think this is important to fix before releasing this.

ThePython10110 commented:
The problem was `level = 1`. In MCL, all tools are level 0, since you can use any tool to break any node (but it doesn't drop unless the tool is good enough).
